src/Vault.sol

```diff
@@ -4,38 +4,152 @@ pragma solidity ^0.8.0;
     event Withdraw(address indexed user, uint256 amount);
+    event BorrowedByMarket(address indexed market, uint256 amount);
+    event RepaidToVault(address indexed market, uint256 amount);
+    event Deposit(address indexed user, uint256 amount, uint256 shares);
+    event Withdraw(address indexed user, uint256 amount, uint256 shares);
+
+    Market public market; // Store the market
+    address public owner;
 
     constructor(
-        address _asset,
+        IERC20 _asset,
+        address _marketContract,
         string memory _name, // name of the vault share token
         string memory _symbol // symbol of the vault share token
-    ) ERC20(_name, _symbol) ERC4626(IERC20(_asset)) {}
+    ) ERC20(_name, _symbol) ERC4626(IERC20(_asset)) {
+        owner = msg.sender;
+        require(address(_asset) != address(0), "Invalid asset address");
+        if (_marketContract != address(0)) {
+            market = Market(_marketContract);
+        }
+    }
+
+    modifier onlyOwner() {
+        require(
+            msg.sender == owner,
+            "Only contract owner can execute this function"
+        );
+        _;
+    }
+
+    modifier onlyMarket() {
+        require(
+            msg.sender == address(market),
+            "Only Market contract can execute this function"
+        );
+        _;
+    }
+
+    modifier onlyValidMarket() {
+        require(address(market) != address(0), "Invalid market address");
+        _;
+    }
 
     /// @notice Deposit ERC-20 tokens into the vault
     function deposit(
-        uint256 assets,
+        uint256 amount,
         address receiver
-    ) public override returns (uint256 shares) {
-        require(assets > 0, " Deposit amount must be greater than 0");
-
-        // Call the ERC4626 deposit logic
-        shares = super.deposit(assets, receiver);
-
+    ) public override nonReentrant returns (uint256 shares) {
+        require(amount > 0, "Deposit amount must be greater than 0");
+        shares = super.deposit(amount, receiver);
+        emit Deposit(receiver, amount, shares);
         return shares;
     }
 
     function withdraw(
-        uint256 assets,
+        uint256 amount,
         address receiver,
-        address owner
-    ) public override returns (uint256 shares) {
-        // Call the default ERC4626 withdraw logic
-        shares = super.withdraw(assets, receiver, owner);
-
+        address user
+    ) public override nonReentrant returns (uint256 shares) {
+        require(amount > 0, "Withdraw amount must be greater than 0");
+        shares = super.withdraw(amount, receiver, user);
+        emit Withdraw(user, amount, shares);
         return shares;
     }
+
+    // Admin function to borrow tokens, only callable by the market contract
+    function adminBorrow(uint256 amount) external nonReentrant onlyMarket {
+        IERC20 token = IERC20(asset());
+        require(address(token) != address(0), "Asset token is not set");
```

Comment:
that case doesnt need to be checked since its impossible to create a vault without the token. This also means you dont need to store token. you can simply call the asset.transfer directly down below